src/PowerShellEditorServices/Console/WindowsConsoleOperations.cs

```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.PowerShell.EditorServices.Console
+{
+    internal class WindowsConsoleOperations : IConsoleOperations
+    {
+        private ConsoleKeyInfo? _bufferedKey;
+
+        private SemaphoreSlim _readKeyHandle = new SemaphoreSlim(1, 1);
+
+        public async Task<ConsoleKeyInfo> ReadKeyAsync(CancellationToken cancellationToken)
+        {
+            await _readKeyHandle.WaitAsync(cancellationToken);
+            try
+            {
+                return
+                    _bufferedKey.HasValue
+                        ? _bufferedKey.Value
+                        : await Task.Factory.StartNew(
+                            () => (_bufferedKey = System.Console.ReadKey(intercept: true)).Value);
```

Comment:
I think this implementation might have a different effect than the original, but correct me if I'm wrong.  The original implementation only stores the buffered key if the cancellation token is cancelled so that it can be returned the next time the ReadKeyAsync method is called.

In the new implementation, we're buffering the key on each read, potentially causing the same key to be returned twice since it gets returned the first time and then buffered to be returned once again?

Reply:
Patrick just explained how this works to me over Slack, the `finally` block takes care of clearing the bufferedKey if the CancellationToken hasn't been cancelled :)  Looks good!